### PR TITLE
Java fixes

### PR DESCRIPTION
**Pull request by bobsmith947**
Various fixes to the logic. The Symbol class now uses a static `of` method to return instances, rather than creating identical instances of rule/alternation/lambda/end of file symbols.

### CODE REVIEW

File: java/compilers/CFG.java [hunk 1]
```diff
@@ -110,7 +111,7 @@ boolean derivesToLambda(Symbol s) {
         List<Symbol> rest = seq.subList(1, seq.size());
 
         if (Symbol.EOF.equals(firstSymbol) || getTerminals().contains(firstSymbol)) {
-            return new HashSet<>(Collections.singletonList(firstSymbol));
+            return hashSet(firstSymbol);
```
Comment (joshdorsey) [inline java/compilers/CFG.java]: This is kind of pedantic, but could you import `SetUtils` instead and make this `SetUtils.hashSet`?  The scope of this function is a little vague otherwise (I thought you had written it at first).

File: java/compilers/CFGReader.java [hunk 2]
```diff
@@ -19,11 +19,11 @@ private static CFG readGrammar(File cfg) {
 
             while (in.hasNext()) {
                 String line = in.nextLine();
-                if(line.trim().isBlank()) {
+                if(line.isBlank()) {
```
Comment (joshdorsey) [inline java/compilers/CFGReader.java:22]: Why'd you pull the trim off of this?  This will cause an NPE on line 34 on lines containing only whitespace.

File: java/compilers/Symbol.java [hunk 3]
```diff
@@ -5,26 +5,48 @@
...
-    public boolean isTerminal() {
-        return SymbolType.TERMINAL.equals(type);
+    boolean isTerminal() {
+        return type == SymbolType.TERMINAL;
     }
 
-    public boolean isNonTerminal() {
-        return SymbolType.NONTERMINAL.equals(type);
+    boolean isNonTerminal() {
+        return type == SymbolType.NONTERMINAL;
```
Comment (joshdorsey) [inline java/compilers/Symbol.java:46]: Why switch the enum comparisons like this?

File: java/compilers/ParseTree.java [hunk 4]
```diff
@@ -33,11 +33,12 @@
 				ParseNode node = new ParseNode(s);
 				cur.addChild(node);
 				cur = node;
-			} else if (s.isTerminal() || s == Symbol.EOF || s == Symbol.LAMBDA) {
-				if (s != Symbol.LAMBDA) {
+			} else if (s.isAugmentedSigma() || s == Symbol.LAMBDA) {
```
Comment (joshdorsey) [inline java/compilers/ParseTree.java:36]: correct me if i'm wrong, but I think `Sigma + EOF + LAMBDA = (Nonterminals)'` so this condition can be written `!s.isNonTerminal()` instead.

File: java/compilers/Symbol.java [hunk 5]
```diff
@@ -33,25 +55,14 @@ public String toString() {
...
-                    return SymbolType.TERMINAL;
-                if (token.matches("[A-Z]+"))
-                    return SymbolType.NONTERMINAL;
-                return null;
-        }
-    }
-
     @Override
     public boolean equals(Object o) {
-        if (this == o) return true;
```
Comment (joshdorsey) [inline java/compilers/Symbol.java:53]: Manually editing `.equals()` from the generated one like this guarantees that the semantics of `.equals()` and `.hashCode()` are out of sync now.  If I'm not mistaken, now `a.equals(b)` may be true when `a.hashCode() != b.hashCode()` in the case that one of these two is a token equal to EOF but not the same instance as `Symbol#EOF`.  You already handled this case when you made the constructor private since the static `Symbol#of` function checks for this, so please revert this change unless there's a reason I'm missing.

File: java/compilers/ParseTree.java [hunk 6]
```diff
@@ -33,11 +33,12 @@
...
+				if (s.isAugmentedSigma()) {
 					if (!s.equals(input.peek()))
 						throw new Exception(input.peek() + ": does not match expected terminal: " + s);
-					input.pop();
+					if (s != Symbol.EOF)
+						input.pop();
```
Comment (joshdorsey) [inline java/compilers/ParseTree.java:41]: I'd want curly brackets around this if-statement unless it was all on one line, but that's just me (that actually wasted about 20 minutes of my time when I first started working on the code because I made a dumb mistake that assumed the brackets would be there).

File: java/compilers/CFGReader.java [hunk 7]
```diff
@@ -69,7 +69,7 @@ public static void main(String... args) {
 		LinkedList<Symbol> lines = new LinkedList<>();
 		try (Scanner in = new Scanner(new File(args[1]))) {
 			while (in.hasNext())
-				lines.add(new Symbol(in.nextLine().trim()));
+				lines.add(Symbol.of(in.nextLine().trim()));
```
Comment (joshdorsey) [inline java/compilers/CFGReader.java:72]: hmm, why are you trimming all of the lines before making the parse tree?

File: java/compilers/CFGReader.java (hunk 2 above)
Comment (bobsmith947) [reply java/compilers/CFGReader.java:22]: I did not see this behavior, unless it is some sort of edge case. If a line contains only whitespace, trimming the leading and trailing whitespace should not affect anything.

File: java/compilers/Symbol.java (hunk 3 above)
Comment (bobsmith947) [reply java/compilers/Symbol.java:46]: Mostly for style, to be more consistent.

File: java/compilers/ParseTree.java (hunk 4 above)
Comment (bobsmith947) [reply java/compilers/ParseTree.java:36]: Yes I believe this would also be valid, but is also slightly vague. The pseudocode expresses it in this way which I think makes the most sense.

File: java/compilers/Symbol.java (hunk 5 above)
Comment (bobsmith947) [reply java/compilers/Symbol.java:53]: The reason behind this is that `Symbol.EOF.equals(null)` will be true, which is desirable since the input queue will return null when all input is consumed. It would technically break the contract between `equals` and `hashCode` since null has no hash code, but helps simplify the parsing.

File: java/compilers/CFGReader.java (hunk 7 above)
Comment (bobsmith947) [reply java/compilers/CFGReader.java:72]: `Symbol#of` will not match anything with whitespace characters, the input token should be lowercase characters only.

File: java/compilers/CFG.java (hunk 1 above)
Comment (bobsmith947) [reply java/compilers/CFG.java]: b916198af97809d73ae570131a8c5cc7a1549c86

File: java/compilers/CFGReader.java (hunk 2 above)
Comment (joshdorsey) [reply java/compilers/CFGReader.java:22]: I was thinking that it would allow a whitespace-only line into this block, and then line 34 would throw.  It looks like `String#isBlank` checks if the line is only whitespace so there isn't an issue.

File: java/compilers/CFGReader.java (hunk 7 above)
Comment (joshdorsey) [reply java/compilers/CFGReader.java:72]: Ah, for some reason I was picturing a python-like scenario where the whitepsace sometimes generates tokens.

File: java/compilers/Symbol.java [hunk 8]
```diff
@@ -5,26 +5,48 @@
...
+    }
+
+    static Symbol of(String token) {
+        switch (token) {
+            case "->": return RULE;
+            case "|": return ALT;
+            case "$": return EOF;
+            case "lambda": return LAMBDA;
+            default:
+                if (token.matches("[a-z]+"))
```
Comment (joshdorsey) [inline java/compilers/Symbol.java:33]: I wonder if using a regex is faster than just doing `token.equals(token.toUpperCase())`

File: java/compilers/Symbol.java (hunk 3 above)
Comment (joshdorsey) [reply java/compilers/Symbol.java:46]: I see that they're all the same now, but that's just because you changed them all from the old style.  I was wondering why you prefer this style specifically.

File: java/compilers/Symbol.java (hunk 5 above)
Comment (joshdorsey) [reply java/compilers/Symbol.java:53]: So this is a solved problem we've discussed in class.  IIRC Hellman mentioned that the input queue is always padded with some number of EOF tokens.  That's the way to solve this problem, not a hack like this.  I'm also not positive that this code gets run the way you're imagining.
There are a couple of places where you are equality comparing instances of the symbol class rather than the `SymbolType` enum (line 61 here is one you added and in ParseTree.java:36 you compare a symbol with Symbol.LAMBDA which is shady for the same reasons). This seems like bad design to me, and I think it will eventually cause a bug.
This is why I prefer using .equals on enums because you're still guaranteed no NPE and it's harder to mix up the semantics between the two types, and totally sidestepping the `Symbol#equals` function.
Maybe in this case you know from the static factory method and the code you've written which uses this class that there will never be any dissonance between reference and identity, maybe you sidestepped the equals logic intentionally in those places, but the members on this class aren't final, and there's nothing to stop the reference and value semantics from getting out of sync.  There's also now an empty constructor that didn't exist before, for a class which I would picture as immutable.
When I picture the Symbol class it has a couple of key requirements:
1) it's immutable.  the class is constructed based on an object that exists in the parsed source code or in our grammar, and because of that can't be changed.
2) it's comparable using equals.  we don't want to have to track the places a special value is created (like EOF) to make sure that our identity comparisons are always to the right instances ones.  (imagine someone just makes an empty Symbol, and populates it with token="$", type=SymbolType.EOF, then ParseTree.java:40 might fail sometimes)
3) it's not nullable.  if it's absolutely necessary there could be a special value for 'empty symbol' or something like that, but if it's possible this should be avoided imo (and so should null).  Ideally I think `Symbol#of` would return an `Optional<Symbol>`, and that way an error constructing a symbol is an explicit absence of a value, not null (which is ambiguous).

tl;dr there better ways to solve this problem, there are ways to get better language-level guarantees of code correctness by constraining this class more, the current version seems error-prone to me based on my java experience on very large projects.

Comment (bobsmith947) [reply java/compilers/Symbol.java:53]: Yes this is valid, it was meant as a simple fix and something I can work on later. Quoted in #6 for future reference.

**Comment by bobsmith947** [comment]
Will be superseded by a future PR for item sets.
